pravega/pravega
# Issue 4874: Update hadoopVersion

**Change log description**  
This touches only one file gradle.properties to update the version of pravega-hdfs.
The hadoopVersion is updated to 2.9.2 as described in the linked issue. However, the guava version had to be lowered slightly. This is not a mistake. All revisions upwards of 2.9.2 up to the latest have had this breaking change which is described by the issues [HDFS-752](https://issues.apache.org/jira/browse/HDFS-752) and [HDFS-7477](https://issues.apache.org/jira/browse/HDFS-7477) and they don't agree with guava. The links for these issues are mentioned here so that the patch associated with these issues can be viewed. They are the only break-in changes we have. These changes are applicable to all recent versions but 2.9.2 as described in the attached issue fixes all the security vulnerabilities cited in the GitHub issue [4874](https://github.com/pravega/pravega/issues/4874) linked to this request.  

**Purpose of the change**  
The purpose of this change is to remove the security vulnerabilities associated with the hadoop version. The Hadoop Version poses a bigger security threat than others to Pravega.

Fixes #4874.

**What the code does**  
This code just changes versions in gradle.properties

**How to verify it**  
This change was verified by travis build.

The exception faced with upgrading Guava is
```
io.pravega.storage.hdfs.HDFSStorageTest > testDelete FAILED
    java.lang.NoSuchMethodError: com.google.common.util.concurrent.Futures.addCallback(Lcom/google/common/util/concurrent/ListenableFuture;Lcom/google/common/util/concurrent/FutureCallback;)V
        at org.apache.hadoop.hdfs.server.datanode.checker.ThrottledAsyncChecker.addResultCachingCallback(ThrottledAsyncChecker.java:167)
        at org.apache.hadoop.hdfs.server.datanode.checker.ThrottledAsyncChecker.schedule(ThrottledAsyncChecker.java:156)
        at org.apache.hadoop.hdfs.server.datanode.checker.StorageLocationChecker.check(StorageLocationChecker.java:164)
        at org.apache.hadoop.hdfs.server.datanode.DataNode.makeInstance(DataNode.java:2762)
        at org.apache.hadoop.hdfs.server.datanode.DataNode.instantiateDataNode(DataNode.java:2677)
        at org.apache.hadoop.hdfs.MiniDFSCluster.startDataNodes(MiniDFSCluster.java:1663)
        at org.apache.hadoop.hdfs.MiniDFSCluster.initMiniDFSCluster(MiniDFSCluster.java:905)
        at org.apache.hadoop.hdfs.MiniDFSCluster.<init>(MiniDFSCluster.java:517)
        at org.apache.hadoop.hdfs.MiniDFSCluster$Builder.build(MiniDFSCluster.java:476)
        at io.pravega.storage.hdfs.HDFSClusterHelpers.createMiniDFSCluster(HDFSClusterHelpers.java:32)
        at io.pravega.storage.hdfs.HDFSStorageTest.setUp(HDFSStorageTest.java:68)
``` 
This method was removed from this library as seen by [others](https://issues.jenkins-ci.org/browse/JENKINS-56252) as well. This [issue](https://issues.apache.org/jira/browse/BEAM-3156) is tracked on the originator side as well with this [jira](https://issues.apache.org/jira/browse/HADOOP-16765). Notice the method takes two parameters instead of the [three](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/util/concurrent/FutureCallback.html) in recent versions. The usual resolution of using exclusions, diff'ing dependency trees and ordering dependency load such as this [one](https://stackoverflow.com/questions/8257545/nosuchmethoderror-exception-when-using-com-google-common-base-splitter/23253320#23253320) do not work.

Testing:
```
docker stack services pravega
ID                  NAME                   MODE                REPLICAS            IMAGE                                                                          PORTS
127hedtw2lzp        pravega_hdfs           replicated          1/1                 pravega/hdfs:2.7.7                                                        
182fk0erog8d        pravega_segmentstore   replicated          1/1                 devops-repo.isus.emc.com:8116/rajamr2/pravega:0.8.0-2577.5d687b9-SNAPSHOT      *:12345->12345/tcp
lcdxxqpdx3h9        pravega_zookeeper      replicated          1/1                 zookeeper:3.5.4-beta                                                      
m9xy1i32zerv        pravega_bookkeeper     replicated          3/3                 devops-repo.isus.emc.com:8116/rajamr2/bookkeeper:0.8.0-2577.5d687b9-SNAPSHOT
y8vrw86ysu7y        pravega_controller     replicated          1/1                 devops-repo.isus.emc.com:8116/rajamr2/pravega:0.8.0-2577.5d687b9-SNAPSHOT      *:9090->9090/tcp,*:10080->10080/tcp

```
Note that the exception described above is being addressed with different jars where the newer one is different from the older one) and the guava blogs also calls out this issue:
https://mvnrepository.com/artifact/com.google.guava/listenablefuture
[Downgrading](https://blog.gradle.org/version-downgrade) is the only option left at this time.
[GuavaVersion29versus25.txt](https://github.com/pravega/pravega/files/4815622/GuavaVersion29versus25.txt)

Unit-testing:
```
404 tests completed, 1 failed
StreamSegmentContainerTests. testWriteFenceOut
org.junit.runners.model.TestTimedOutException: test timed out after 100000 milliseconds
#  rerunning
io.pravega.segmentstore.server.containers.StreamSegmentContainerTests > testWriteFenceOut PASSED
```


Testing with 3.1.1:
```
docker stack services pravega
ID                  NAME                   MODE                REPLICAS            IMAGE                                                                          PORTS
c0ivwwzvonrs        pravega_zookeeper      replicated          1/1                 zookeeper:3.5.4-beta                                                      
dgolahviacf7        pravega_hdfs           replicated          1/1                 devops-repo.isus.emc.com:8116/rajamr2/hdfs:3.1.1                                                       
n5114asbhf90        pravega_controller     replicated          1/1                 devops-repo.isus.emc.com:8116/rajamr2/pravega:0.8.0-2577.5d687b9-SNAPSHOT      *:9090->9090/tcp,*:10080->10080/tcp
o5yz5hmaan8d        pravega_bookkeeper     replicated          3/3                 devops-repo.isus.emc.com:8116/rajamr2/bookkeeper:0.8.0-2577.5d687b9-SNAPSHOT
qo9ushm24t2f        pravega_segmentstore   replicated          1/1                 devops-repo.isus.emc.com:8116/rajamr2/pravega:0.8.0-2577.5d687b9-SNAPSHOT      *:12345->12345/tcp

```
